### ISSUE TITLE
varnish_backend_up always 0

### ISSUE THREAD

**Issue by strowi** [jonnenauha/prometheus_varnish_exporter#23]
Hi,

we're using a rather old varnish-3.0.7, and all the varnish_backend_up metrics are 0...
I'm guessing the value is somehow calculated from the happy metrics? Could you be so kind and elaborate on that e.g. how can i calculate the value instead?

greetings,
strowi

**Comment by jonnenauha** [comment]
Yes, its calculated from the happy metric from varnishstat. They call it a bitmap value. From what I gathered from their sources is its and uint64 value where it looks to be storing 24 values.

Here is what I wrote https://github.com/jonnenauha/prometheus_varnish_exporter/blob/master/varnish.go#L142-L164 
Here is where I read how the bitmap works https://github.com/varnishcache/varnish-cache/blob/master/bin/varnishstat/varnishstat_curses.c#L654-L694 

I could not find any documentation what this bitmap is and how it should be read, that was something I could follow. Comparing that code (24 'V' or '_'). Below is what varnishstat renders.

```
VBE.boot.eu1.happy    ffffffffff     VVVVVVVVVVVVVVVVVVVVVVVV
VBE.boot.us1.happy    ffffffffff     VVVVVVVVVVVVVVVVVVVVVVVV
```

Do you have healt probes enabled? I'm sure the happy value is still reported by varnishstat, but as zero. Which would make it always zero. If you have health probes then it could just be your 3.0.7 the happy is constructed in a different manner. We can go back in varnish repo to the release and look.

What do you see in `varnishstat` and what is the number value of that metric if you run `varnishstat -j` or `-1`. For example I get `18446744073709551615` when all probes are happy (been running a long time).

**Comment by jonnenauha** [comment]
Yeah does seem that the happy is not rendered at least if the `-1` mode. https://github.com/varnishcache/varnish-cache/blob/varnish-3.0.7/bin/varnishstat/varnishstat_curses.c 

But if there is no metric ending with "_happy", you should not be seeing the "_up" either. As you can see from my commit. So what is the value of happy in your version?

**Comment by strowi** [comment]
Thx for the very elaborate answer.. i will check the details tonight.
For now i have lots of backends like this showing up in varnishstat -1:
```
.alexa_images_core04(10.10.0.110,,30054).happy         18446744073709551615          .   Happy health probes
```

**Comment by jonnenauha** [comment]
Ok that is quite strange. The value is exatly the same than my metrics (with varnish 5.x). I cant think of why it would produce 0 for you but 1 for me.

The _happy value is not very useful actually. Prometheus only accepts float64 values. This metric has significance in the max value of uin64 which wont show correctly as a float64, it will be rounded. I should probably drop the happy values all together, I doubt anyone has built valuable and reliable queries with them. And we should now have this replacement.

Actually. One thing I could think of is JSON schema changes. My code detects that the type must be "b" for bitmap. If it does not report that with 3.x then it will definitely set a zero there. Could you please provide me with `varnishstat -j` output for one full happy section. You can obfuscate any sensitive info as you did before, those are not important.

**Comment by strowi** [comment]
that could make sense... looks like there are some different naming schemes..

```
  "VBE.alexa_images_core03(10.10.0.104,,30054).vcls": {"type": "VBE", "ident": "alexa_images_core03(10.10.0.104,,30054)", "value": 3, "flag": "i", "description": "VCL references"},
  "VBE.alexa_images_core03(10.10.0.104,,30054).happy": {"type": "VBE", "ident": "alexa_images_core03(10.10.0.104,,30054)", "value": 18446744073709551615, "flag": "b", "description": "Happy health probes"},
  "VBE.alexa_images_core04(10.10.0.110,,30054).vcls": {"type": "VBE", "ident": "alexa_images_core04(10.10.0.110,,30054)", "value": 3, "flag": "i", "description": "VCL references"},
  "VBE.alexa_images_core04(10.10.0.110,,30054).happy": {"type": "VBE", "ident": "alexa_images_core04(10.10.0.110,,30054)", "value": 18446744073709551615, "flag": "b", "description": "Happy health probes"},
  "VBE.alexa_images_core05(10.10.0.109,,30054).vcls": {"type": "VBE", "ident": "alexa_images_core05(10.10.0.109,,30054)", "value": 3, "flag": "i", "description": "VCL references"},
  "VBE.alexa_images_core05(10.10.0.109,,30054).happy": {"type": "VBE", "ident": "alexa_images_core05(10.10.0.109,,30054)", "value": 18446744073709551615, "flag": "b", "description": "Happy health probes"},
  "VBE.alexa_images_core06(10.10.0.108,,30054).vcls": {"type": "VBE", "ident": "alexa_images_core06(10.10.0.108,,30054)", "value": 3, "flag": "i", "description": "VCL references"},
  "VBE.alexa_images_core06(10.10.0.108,,30054).happy": {"type": "VBE", "ident": "alexa_images_core06(10.10.0.108,,30054)", "value": 18446744073709551615, "flag": "b", "description": "Happy health probes"},
```

**Comment by jonnenauha** [comment]
Ah yes, the field there is "flag" and in new versions its "format". I'll fix this asap when I figure out what version changed "flag" into "format" so I can do a proper fix.

Here is where it fails https://github.com/jonnenauha/prometheus_varnish_exporter/blob/master/varnish.go#L95-L116

**Comment by jonnenauha** [comment]
Ok, looks like it is introduced in 4.1. I try to cook this up in the evening.

https://varnish-cache.org/docs/4.0/reference/varnishstat.html 
vs
https://varnish-cache.org/docs/4.1/reference/varnishstat.html 
(code https://github.com/varnishcache/varnish-cache/blob/4.1/bin/varnishstat/varnishstat.c#L130)

**Comment by strowi** [comment]
That'd be great, but no rush... ;)

**Comment by jonnenauha** [comment]
@strowi any change you could build from sources and test? I pushed a commit that should resolve this.

**Comment by strowi** [comment]
Hi,

looks good to me.. (the '0'-ones are without a probe).

```
varnish_backend_up{backend="metrics_bay_server_core05",server="10.10.0.109:30069"} 0
varnish_backend_up{backend="metrics_bay_server_core06",server="10.10.0.108:30069"} 0
varnish_backend_up{backend="mobile_g_micro_server_core03",server="10.10.0.104:30039"} 1
varnish_backend_up{backend="mobile_g_micro_server_core04",server="10.10.0.110:30039"} 1
```
thx for the very fast response + your work! ;)

**Comment by jonnenauha** [comment]
Cool, thanks for reporting 👍

**Comment by jonnenauha** [comment]
@strowi FYI: This feature was just released in 1.4 release. There are no other major changes except prometeus internal metrics not exported by default, so not that necessary to update from your previous manual build.
